crates/typst/src/util/str.rs

```diff
@@ -0,0 +1,63 @@
+use ecow::EcoString;
+use lasso::{Spur, ThreadedRodeo};
+use once_cell::sync::Lazy;
+use typst_macros::cast;
+
+/// The global string interner.
+static INTERNER: Lazy<ThreadedRodeo> = Lazy::new(ThreadedRodeo::new);
+
+/// An interned string.
+///
+/// The API is purposefully kept small. This is because it might be relatively
+/// slow to look up a string in the interner, so we want to avoid doing it
+/// unnecessarily. For this reason, the user should use the [`PicoStr::resolve`]
+/// method to get the underlying string, such that the lookup is done only once.
+#[derive(Clone, Copy, PartialEq, Eq, Ord, PartialOrd, Hash)]
+pub struct PicoStr(Spur);
+
+impl PicoStr {
+    /// Creates a new interned string.
+    pub fn new<S: AsRef<str>>(s: S) -> Self {
+        Self(INTERNER.get_or_intern(s.as_ref()))
+    }
+
+    /// Creates a new interned string from a static string.
+    pub fn static_(s: &'static str) -> Self {
+        Self(INTERNER.get_or_intern_static(s))
+    }
+
+    /// Resolves the interned string.
+    pub fn resolve(&self) -> &'static str {
```

Comment:
How about `as_str`?

Reply:
I was to make it clear that this performs work, I feel like `as_str` while being very concise don't carry the implication of the (surprisingly high) cost of querying a string. The big advantage of `PicoStr` is memory usage, hashing speed, and string compare speed, but at the cost of making the actual "resolving" of the string more costly.

Reply:
Okay